store/engine/postgresql/postgresql.go

```diff
@@ -0,0 +1,299 @@
+
+	leaderMu  sync.Mutex
+	leaderID  string
+	myID      string
+	electPath string
+	isReady   atomic.Bool
+
+	quitCh         chan struct{}
+	wg             sync.WaitGroup
+	lockReleaseCh  chan bool
+	leaderChangeCh chan bool
+}
+
+func New(id string, cfg *Config) (*Postgresql, error) {
+	if len(id) == 0 {
+		return nil, errors.New("id must NOT be a empty string")
+	}
+
+	connStr := fmt.Sprintf("postgres://%s:%s@%s/%s?sslmode=disable", cfg.Username, cfg.Password, cfg.Addrs[0], cfg.DBName)
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		return nil, err
+	}
+
+	listener := pq.NewListener(connStr, listenerMinReconnectInterval, listenerMaxReconnectInterval, nil)
+	err = listener.Listen(cfg.NotifyChannel)
+	if err != nil {
+		return nil, err
+	}
+
+	electPath := defaultElectPath
+	if cfg.ElectPath != "" {
+		electPath = defaultElectPath
+	}
+
+	p := &Postgresql{
+		myID:           id,
+		electPath:      electPath,
+		db:             db,
+		listener:       listener,
+		quitCh:         make(chan struct{}),
+		lockReleaseCh:  make(chan bool),
+		leaderChangeCh: make(chan bool),
+	}
+	err = p.initLeaderId()
+	if err != nil {
+		return nil, err
+	}
+	p.isReady.Store(false)
+	p.wg.Add(2)
+	go p.electLoop()
+	go p.observeLeaderEvent()
+	return p, nil
+}
+
+func (p *Postgresql) ID() string {
+	return p.myID
+}
+
+func (p *Postgresql) Leader() string {
+	p.leaderMu.Lock()
+	defer p.leaderMu.Unlock()
+	return p.leaderID
+}
+
+func (p *Postgresql) LeaderChange() <-chan bool {
+	return p.leaderChangeCh
+}
+
+func (p *Postgresql) IsReady(ctx context.Context) bool {
+	for {
+		select {
+		case <-p.quitCh:
+			return false
+		case <-time.After(100 * time.Millisecond):
+			if p.isReady.Load() {
+				return true
+			}
+		case <-ctx.Done():
+			return p.isReady.Load()
+		}
+	}
+}
+
+func (p *Postgresql) Get(ctx context.Context, key string) ([]byte, error) {
+	var value []byte
+	query := "SELECT value FROM kv WHERE key = $1"
+
+	row := p.db.QueryRow(query, key)
+	err := row.Scan(&value)
+	if err == sql.ErrNoRows {
+		return nil, consts.ErrNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return value, nil
+}
+
+func (p *Postgresql) Exists(ctx context.Context, key string) (bool, error) {
+	_, err := p.Get(ctx, key)
+	if err != nil {
+		if errors.Is(err, consts.ErrNotFound) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (p *Postgresql) Set(ctx context.Context, key string, value []byte) error {
+	query := "INSERT INTO kv (key, value) VALUES ($1, $2) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value"
+	_, err := p.db.Exec(query, key, value)
+	return err
+}
+
+func (p *Postgresql) Delete(ctx context.Context, key string) error {
+	query := "DELETE FROM kv WHERE key = $1"
+	_, err := p.db.Exec(query, key)
+	return err
+}
+
+func (p *Postgresql) List(ctx context.Context, prefix string) ([]engine.Entry, error) {
+	prefixWithWildcard := prefix + "%"
+	query := "SELECT * from kv WHERE key LIKE $1"
```

Comment:
It's better to list the desired columns instead of `*` to make the code more robust.